swarm/src/behaviour/peer_addresses.rs

```diff
@@ -0,0 +1,126 @@
+use crate::{behaviour::FromSwarm, NewExternalAddrOfPeer};
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+use std::collections::HashMap;
+
+/// Struct for tracking peers' external addresses of the [`Swarm`](crate::Swarm).
+#[derive(Debug, Clone, Default)]
+pub struct PeerAddresses(HashMap<PeerId, Vec<Multiaddr>>);
+
+impl PeerAddresses {
+    /// Feed a [`FromSwarm`] event to this struct.
+    ///
+    /// Returns whether the event changed peer's known external addresses.
+    pub fn on_swarm_event(&mut self, event: &FromSwarm) -> bool {
+        if let FromSwarm::NewExternalAddrOfPeer(NewExternalAddrOfPeer { peer_id, addr }) = event {
+            match self.0.get_mut(peer_id) {
+                None => {
+                    let mut empty_vec: Vec<Multiaddr> = Vec::new();
+                    Self::push(&mut empty_vec, addr);
+
+                    self.0.insert(*peer_id, empty_vec);
+                    return true;
+                }
+                Some(addrs) => {
+                    if addrs.contains(addr) {
+                        return false;
+                    }
+                    Self::push(addrs, addr);
+
+                    return true;
+                }
+            }
+        }
+        false
+    }
```

Comment:
@thomaseizinger @mxinden On which event should the address (or all peer's addresses) should be removed? Connection closed? :thinking:

Reply:
Also - wouldn't it just be better to use `PeerCache` from identify?

Reply:
Yes, This definitely needs to be bounded in size using an LRU cache makes sense. I think we can pretty much move the `PeerCache` from identify to here and use this within `identify`.

Reply:
> @thomaseizinger @mxinden On which event should the address (or all peer's addresses) should be removed? Connection closed? 🤔

Part of what we want to achieve with this cache is that you can **re**-connect to a peer after the connection has been closed.

But we can remove it on `DialFailure` because clearly, the address doesn't work! Although we should check on _why_ the dial failed and e.g. don't remove it if it is just because it was `Denied`.